src/features/auth/signup/CompleteSignupForm.tsx

```diff
@@ -0,0 +1,177 @@
+      if (urlToken) {
+        setLoading(true);
+        try {
+          setValue("email", await getSignupEmail(urlToken), {
+            shouldDirty: true,
+          });
+        } catch (err) {
+          dispatch(authError(err.message));
+          history.push(signupRoute);
+          return;
+        }
+        setLoading(false);
+      }
+    })();
+  }, [urlToken, dispatch, location.pathname, setValue, history]);
+
+  const completeSignup = handleSubmit(async (data: SignupInputs) => {
+    const parsedBirthdate = new Date(data.birthdate);
+    dispatch(
+      signup({
+        signupToken: urlToken,
+        username: data.username,
+        name: data.name,
+        city: data.city,
+        birthdate: parsedBirthdate,
+        gender: data.gender,
+        hostingStatus: data.hostingStatus,
+      })
+    );
+  });
+
+  return (
+    <>
+      {loading ? (
+        <CircularProgress />
+      ) : (
+        <form onSubmit={completeSignup}>
+          <Typography variant="h3">{getValues("email")}</Typography>
+          <TextInput
+            name="name"
+            label="Name"
+            inputRef={register({
+              required: "Enter your name",
+              pattern: {
+                value: nameValidationPattern,
+                message: "Name can't be just white space.",
+              },
+            })}
+            helperText={errors?.name?.message}
+          />
+          <TextInput
+            name="username"
+            label="Username"
+            inputRef={register({
+              required: "Enter your username",
+              pattern: {
+                //copied from backend, added ^ at the start
+                value: usernameValidationPattern,
+                message:
+                  "Username can only have lowercase letters, numbers or _, starting with a letter.",
+              },
+              validate: async (username) => {
+                const valid = await validateUsername(username);
+                return valid || "This username is taken.";
+              },
+            })}
+            helperText={errors?.username?.message}
+          />
+          <TextInput
+            name="city"
+            label="City"
+            inputRef={register({
+              required: "Enter your city",
+            })}
+            helperText={errors?.city?.message}
+          />
+          <Autocomplete
+            label="Gender"
+            onInputChange={(_event, value) => setValue("gender", value)}
+            options={["Male", "Female", "<Type anything you like>"]}
+            freeSolo
+          />
+          <TextInput
+            name="birthdate"
+            label="Birthdate"
+            type="date"
+            InputLabelProps={{
+              shrink: true,
+            }}
+            inputRef={register({
+              required: "Enter your birthdate",
+              validate: (stringDate) =>
+                validateBirthdate(stringDate) || "Not a valid date.",
```

Comment:
```suggestion
              validate: validateBirthdate,
```
I think you can do that and move the return value logic into the validation function to clean this up even more?

Reply:
Not sure about that. If a boolean is returned you can use that for further logic, but s string is for those humans understanding the used language.

Reply:
I think I prefer as is, I'm going to leave it unless you feel strongly about it